File: backend/src/application/use-cases/CreateHelpOffer.ts
import { HelpOffer } from '../../domain/entities/HelpOffer';
import { HelpRequestType } from '../../domain/entities/HelpRequest';
import { IHelpOfferRepository } from '../../domain/repositories/IHelpOfferRepository';
import { IEmergencyRepository } from '../../domain/repositories/IEmergencyRepository';

export interface CreateHelpOfferDTO {
  emergencyId: string;
  volunteerId: string;
  type: HelpRequestType;
  description: string;
  latitude: number;
  longitude: number;
  capacity?: number;
}

export class CreateHelpOffer {
  constructor(
    private readonly helpOfferRepository: IHelpOfferRepository,
    private readonly emergencyRepository: IEmergencyRepository
  ) {}

  async execute(dto: CreateHelpOfferDTO): Promise<HelpOffer> {
    const emergency = await this.emergencyRepository.findById(dto.emergencyId);
    if (!emergency) {
      throw new Error('Emergency not found');
    }
    if (!emergency.isActive()) {
      throw new Error('Emergency is no longer active');
    }

    const offer = new HelpOffer({
      emergencyId: dto.emergencyId,
      volunteerId: dto.volunteerId,
      type: dto.type,
      description: dto.description,
      latitude: dto.latitude,
      longitude: dto.longitude,
      capacity: dto.capacity,
    });

    return this.helpOfferRepository.save(offer);
  }
}

File: backend-dotnet/src/Hestia.Application/UseCases/CreateHelpOffer.cs
using Hestia.Application.DTOs;
using Hestia.Domain.Entities;
using Hestia.Domain.Repositories;

namespace Hestia.Application.UseCases;

public class CreateHelpOffer
{
    private readonly IHelpOfferRepository _helpOfferRepository;
    private readonly IEmergencyRepository _emergencyRepository;

    public CreateHelpOffer(IHelpOfferRepository helpOfferRepository, IEmergencyRepository emergencyRepository)
    {
        _helpOfferRepository = helpOfferRepository;
        _emergencyRepository = emergencyRepository;
    }

    public async Task<HelpOffer> ExecuteAsync(CreateHelpOfferRequest dto, Guid volunteerId, CancellationToken ct = default)
    {
        var emergency = await _emergencyRepository.FindByIdAsync(dto.EmergencyId, ct)
            ?? throw new ApplicationException("Emergency not found");

        if (!emergency.IsActive())
            throw new ApplicationException("Emergency is no longer active");

        var offer = new HelpOffer(
            dto.EmergencyId, volunteerId, dto.Type,
            dto.Description, dto.Latitude, dto.Longitude,
            dto.Capacity ?? 1);

        return await _helpOfferRepository.SaveAsync(offer, ct);
    }
}
